File: src/systems/ItemSystem.js
export class ItemSystem {
    constructor(game) {
        this.game = game;
    }
    
    splitItem(item, amount) {
        if (!item.quantity || !item.quantityUnit) {
            return null;
        }
        
        if (amount >= item.quantity) {
            return null;
        }
        
        const newItem = { ...item };
        newItem.id = `${item.name}_${Date.now()}_${Math.random()}`;
        newItem.quantity = amount;
        
        if (item.state) {
            newItem.state = { ...item.state };
        }
        
        item.quantity -= amount;
        
        const weightPerUnit = item.weight / (item.quantity + amount);
        const volumePerUnit = item.volume / (item.quantity + amount);
        
        item.weight = Math.floor(weightPerUnit * item.quantity);
        item.volume = Math.floor(volumePerUnit * item.quantity);
        newItem.weight = Math.floor(weightPerUnit * amount);
        newItem.volume = Math.floor(volumePerUnit * amount);
        
        return newItem;
    }
    
    getAvailableOpeningTools(player, container) {
        if (!container.openMethods) return [];
        
        const tools = [];
        
        for (const [toolType, method] of Object.entries(container.openMethods)) {
            if (toolType === 'hand') {
                tools.push({
                    type: 'hand',
                    name: 'Your Hands',
                    item: null,
                    yield: method.yield,
                    durabilityDamage: method.durabilityDamage
                });
                continue;
            }
            
            if (toolType === 'ground') {
                tools.push({
                    type: 'ground',
                    name: 'Smash on Ground',
                    item: null,
                    yield: method.yield,
                    durabilityDamage: method.durabilityDamage
                });
                continue;
            }
            
            const equippedTool = this.findToolInEquipment(player, toolType);
            if (equippedTool) {
                tools.push({
                    type: toolType,
                    name: equippedTool.name,
                    item: equippedTool,
                    yield: method.yield,
                    durabilityDamage: method.durabilityDamage,
                    location: 'equipped'
                });
                continue;
            }
            
            const inventoryTool = this.findToolInInventory(player, toolType);
            if (inventoryTool) {
                tools.push({
                    type: toolType,
                    name: inventoryTool.name,
                    item: inventoryTool,
                    yield: method.yield,
                    durabilityDamage: method.durabilityDamage,
                    location: 'inventory'
                });
            }
        }
        
        return tools;
    }
    
    findToolInEquipment(player, toolType) {
        for (const slot in player.equipment) {
            const item = player.equipment[slot];
            if (!item) continue;
            
            if (toolType === 'can_opener' && item.tags && item.tags.includes('opener')) {
                return item;
            }
            if (toolType === 'knife' && item.tags && item.tags.includes('sharp')) {
                return item;
            }
            if (toolType === 'pipe' && item.tags && item.tags.includes('blunt')) {
                return item;
            }
        }
        return null;
    }
    
    findToolInInventory(player, toolType) {
        for (const item of player.inventory) {
            if (toolType === 'can_opener' && item.tags && item.tags.includes('opener')) {
                return item;
            }
            if (toolType === 'knife' && item.tags && item.tags.includes('sharp')) {
                return item;
            }
            if (toolType === 'pipe' && item.tags && item.tags.includes('blunt')) {
                return item;
            }
        }
        return null;
    }
    
    openContainer(container, tool, player) {
        if (!container.state || container.state.opened) {
            return { success: false, message: 'Already opened.' };
        }
        
        if (!container.openMethods || !container.openMethods[tool.type]) {
            return { success: false, message: 'Cannot open with this tool.' };
        }
        
        const method = container.openMethods[tool.type];
        
        // Ensure state object exists
        if (!container.state) {
            container.state = {};
        }
        
        container.state.opened = true;
        container.state.sealed = false;
        
        // Update container name to show opened status
        if (!container.name.includes('[OPENED]')) {
            container.name = container.name.replace('Sealed', 'Opened');
        }
        
        if (tool.item && method.durabilityDamage > 0) {
            tool.item.durability = Math.max(0, tool.item.durability - method.durabilityDamage);
            if (tool.item.durability <= 0) {
                this.game.ui.log(`Your ${tool.item.name} broke!`, 'warning');
            }
        }
        
        const spilledItems = [];
        if (method.yield < 1.0 && container.contents.length > 0) {
            for (const content of container.contents) {
                if (content.quantity) {
                    const spillAmount = Math.floor(content.quantity * (1 - method.yield));
                    if (spillAmount > 0) {
                        const spilledItem = this.splitItem(content, spillAmount);
                        if (spilledItem) {
                            spilledItem.state = { contaminated: true };
                            spilledItem.x = player.x;
                            spilledItem.y = player.y;
                            spilledItems.push(spilledItem);
                        }
                    }
                }
            }
        }
        
        return {
            success: true,
            message: `Opened ${container.name} with ${tool.name}.`,
            spilledItems: spilledItems,
            yield: method.yield
        };
    }
    
    calculateOptimalConsumption(item, player) {
        if (!item.quantity || !item.nutrition) {
            return item.quantity || 1;
        }
        
        const hungerNeeded = player.maxHunger - player.hunger;
        const thirstNeeded = player.maxThirst - player.thirst;
        
        const hungerPerUnit = item.nutrition.hunger / item.quantity;
        const thirstPerUnit = item.nutrition.thirst / item.quantity;
        
        let amountForHunger = Infinity;
        let amountForThirst = Infinity;
        
        if (hungerPerUnit > 0 && hungerNeeded > 0) {
            amountForHunger = Math.ceil(hungerNeeded / hungerPerUnit);
        }
        if (thirstPerUnit > 0 && thirstNeeded > 0) {
            amountForThirst = Math.ceil(thirstNeeded / thirstPerUnit);
        }
        
        const optimalAmount = Math.min(
            item.quantity,
            amountForHunger,
            amountForThirst
        );
        
        // If player is full or calculation resulted in invalid amount, consume all
        if (optimalAmount === Infinity || optimalAmount <= 0) {
            return item.quantity;
        }
        
        return optimalAmount;
    }
    
    consumeFood(item, player, amount = null) {
        if (!item.nutrition) {
            return { success: false, message: 'This is not consumable.' };
        }
        
        // Check if item has quantity and it's 0 or less
        if (item.quantity !== undefined && item.quantity <= 0) {
            return { success: false, message: 'Nothing left to consume.' };
        }
        
        // Smart consumption: calculate optimal amount if not specified
        let consumeAmount = amount;
        if (consumeAmount === null && item.quantity) {
            consumeAmount = this.calculateOptimalConsumption(item, player);
        } else if (consumeAmount === null) {
            consumeAmount = item.quantity || 1;
        }
        
        if (item.quantity && consumeAmount > item.quantity) {
            return { success: false, message: 'Not enough to consume.' };
        }
        
        // Calculate ratio - if no quantity system, consume all (ratio 1.0)
        // If quantity exists, calculate based on amount consumed
        const ratio = item.quantity ? (consumeAmount / item.quantity) : 1.0;
        
        if (item.nutrition.hunger) {
            player.hunger = Math.min(player.maxHunger, player.hunger + (item.nutrition.hunger * ratio));
        }
        if (item.nutrition.thirst) {
            player.thirst = Math.min(player.maxThirst, player.thirst + (item.nutrition.thirst * ratio));
        }
        
        if (item.state && item.state.contaminated) {
            let contaminationLevel = 0.3;
            
            // Apply ironStomach trait resistance
            if (player.traitEffects && player.traitEffects.poisonResist) {
                contaminationLevel *= (1 - player.traitEffects.poisonResist);
            }
            
            // Only apply sickness if contamination level is significant
            if (contaminationLevel > 0.05) {
                const sicknessDuration = Math.floor(10 * contaminationLevel);
                const sicknessValue = Math.floor(-1 * contaminationLevel);
                
                player.addStatusEffect({
                    type: 'sickness',
                    value: sicknessValue,
                    duration: sicknessDuration,
                    name: 'Food Poisoning'
                });
            }
        }
        
        if (item.quantity) {
            item.quantity -= consumeAmount;
            
            const weightPerUnit = item.weight / (item.quantity + consumeAmount);
            const volumePerUnit = item.volume / (item.quantity + consumeAmount);
            item.weight = Math.floor(weightPerUnit * item.quantity);
            item.volume = Math.floor(volumePerUnit * item.quantity);
        }
        
        return {
            success: true,
            consumed: consumeAmount,
            remaining: item.quantity || 0,
            contaminated: item.state && item.state.contaminated
        };
    }
}

File: Nightrun_v01/Systems/ItemSystem.cs
using Nightrun.Content;
using Nightrun.Entities;

namespace Nightrun.Systems;

/// <summary>
/// A single way the player can open a container — wraps the JS openMethods
/// entry plus the specific tool item (or null for hand/ground methods).
/// </summary>
public sealed class OpenOption
{
    public required string ToolType;        // "can_opener", "knife", "pipe", "hand", "ground"
    public required string DisplayName;     // "Can Opener", "Your Hands", "Smash on Ground", ...
    public Item? ToolItem;                  // null for hand/ground
    public required double Yield;
    public required int DurabilityDamage;
    public string Location = "";            // "equipped" | "inventory" | "" (hand/ground)
}

public readonly record struct OpenResult(
    bool   Success,
    string Message,
    double Yield,
    List<Item> SpilledItems);

public readonly record struct ConsumeResult(
    bool   Success,
    string Message,
    int    Consumed,
    int    Remaining,
    bool   Contaminated);

/// <summary>
/// Port of <c>src/systems/ItemSystem.js</c> — deals with opening sealed
/// containers, finding tools for that, and consuming food/drink. Keeps the
/// same behavioral rules: yield/durability tradeoffs, contamination spill,
/// quantity-based consumption, sickness on contaminated items.
/// </summary>
public sealed class ItemSystem
{
    /// <summary>
    /// Enumerate every way the player can open <paramref name="container"/>
    /// given their current equipment + inventory. The UI presents this list.
    /// Mirrors <c>getAvailableOpeningTools(player, container)</c>.
    /// </summary>
    public static List<OpenOption> GetAvailableOpeningTools(Player player, Item container)
    {
        var options = new List<OpenOption>();
        if (container.OpenMethods == null) return options;

        foreach (var (toolType, method) in container.OpenMethods)
        {
            switch (toolType)
            {
                case "hand":
                    options.Add(new OpenOption
                    {
                        ToolType = "hand", DisplayName = "Your Hands",
                        ToolItem = null, Yield = method.Yield, DurabilityDamage = method.DurabilityDamage,
                    });
                    continue;

                case "ground":
                    options.Add(new OpenOption
                    {
                        ToolType = "ground", DisplayName = "Smash on Ground",
                        ToolItem = null, Yield = method.Yield, DurabilityDamage = method.DurabilityDamage,
                    });
                    continue;
            }

            // Otherwise: look for a tool item with the matching tag, prefer equipped.
            var tool = FindToolInEquipment(player, toolType)
                    ?? FindToolInInventory(player, toolType);
            if (tool != null)
            {
                options.Add(new OpenOption
                {
                    ToolType = toolType, DisplayName = tool.Name,
                    ToolItem = tool, Yield = method.Yield, DurabilityDamage = method.DurabilityDamage,
                    Location = IsEquipped(player, tool) ? "equipped" : "inventory",
                });
            }
        }

        return options;
    }

    private static bool IsEquipped(Player p, Item tool)
    {
        foreach (var (_, it) in p.Equipment.All())
            if (ReferenceEquals(it, tool)) return true;
        return false;
    }

    private static Item? FindToolInEquipment(Player p, string toolType)
    {
        foreach (var (_, item) in p.Equipment.All())
            if (MatchesTool(item, toolType)) return item;
        return null;
    }

    private static Item? FindToolInInventory(Player p, string toolType)
    {
        foreach (var pr in p.Inventory.Pockets())
            foreach (var item in pr.Pocket.Contents)
                if (MatchesTool(item, toolType)) return item;
        return null;
    }

    private static bool MatchesTool(Item item, string toolType) => toolType switch
    {
        "can_opener" => item.HasTag("opener"),
        "knife"      => item.HasTag("sharp"),
        "pipe"       => item.HasTag("blunt"),
        _            => false,
    };

    // ------------------------------------------------------------------
    // Opening

    /// <summary>
    /// Apply <paramref name="option"/> to <paramref name="container"/>.
    /// Flips the container's state to opened, damages the tool if applicable,
    /// and if yield &lt; 1.0 spills a fraction of each stackable content as
    /// contaminated waste (returned in the result for the caller to drop).
    /// </summary>
    public static OpenResult OpenContainer(Item container, OpenOption option, Player player)
    {
        var spilled = new List<Item>();

        if (container.State.Opened)
            return new OpenResult(false, "Already opened.", option.Yield, spilled);

        if (container.OpenMethods == null || !container.OpenMethods.ContainsKey(option.ToolType))
            return new OpenResult(false, "Cannot open with this tool.", option.Yield, spilled);

        // Flip state
        container.State.Opened = true;
        container.State.Sealed = false;

        // Rename: "Sealed Can" → "Opened Can"
        if (container.Name.StartsWith("Sealed", StringComparison.Ordinal))
            container.Name = "Opened" + container.Name["Sealed".Length..];

        // Tool durability damage
        if (option.ToolItem != null && option.DurabilityDamage > 0)
        {
            option.ToolItem.Durability = Math.Max(0, option.ToolItem.Durability - option.DurabilityDamage);
        }

        // Partial yield → spill the lost fraction of each stackable content
        if (option.Yield < 1.0)
        {
            foreach (var pocket in container.Pockets)
            {
                // Iterate a snapshot because we mutate Contents via split
                var contentSnapshot = pocket.Contents.ToArray();
                foreach (var content in contentSnapshot)
                {
                    if (content.Quantity <= 1) continue;  // discrete items don't split
                    int spillAmount = (int)Math.Floor(content.Quantity * (1.0 - option.Yield));
                    if (spillAmount <= 0) continue;
                    var split = SplitItem(pocket, content, spillAmount);
                    if (split != null)
                    {
                        split.State.Contaminated = true;
                        spilled.Add(split);
                    }
                }
            }
        }

        return new OpenResult(true, $"Opened {container.Name} with {option.DisplayName}.",
                              option.Yield, spilled);
    }

    /// <summary>
    /// Split a quantity-based item — reduces the original's quantity/weight/volume
    /// and returns a new Item with the split-off portion. Null if the split is
    /// invalid (not stackable or amount &gt;= quantity). Mirrors splitItem() in JS.
    /// </summary>
    public static Item? SplitItem(Pocket? sourcePocket, Item item, int amount)
    {
        if (item.Quantity <= 0 || string.IsNullOrEmpty(item.QuantityUnit)) return null;
        if (amount >= item.Quantity) return null;

        var split = item.Clone();
        split.Quantity = amount;

        int totalQ = item.Quantity;
        double wPerUnit = item.Weight / (double)totalQ;
        double vPerUnit = item.Volume / (double)totalQ;

        item.Quantity -= amount;
        // Rebalance weight/volume on both items proportionally
        int newOrigW = (int)Math.Floor(wPerUnit * item.Quantity);
        int newOrigV = (int)Math.Floor(vPerUnit * item.Quantity);
        int newSplitW = (int)Math.Floor(wPerUnit * amount);
        int newSplitV = (int)Math.Floor(vPerUnit * amount);

        // If the item lives in a pocket, reflect the weight/volume delta there.
        if (sourcePocket != null)
        {
            sourcePocket.UsedWeight += newOrigW - item.Weight;
            sourcePocket.UsedVolume += newOrigV - item.Volume;
        }

        item.Weight = newOrigW;
        item.Volume = newOrigV;
        split.Weight = newSplitW;
        split.Volume = newSplitV;

        return split;
    }

    // ------------------------------------------------------------------
    // Consumption

    /// <summary>
    /// Calculate how much of <paramref name="item"/> to consume to fill
    /// the player without overshooting. Mirrors <c>calculateOptimalConsumption</c>.
    /// </summary>
    public static int CalculateOptimalConsumption(Item item, Player player)
    {
        if (item.Nutrition == null) return Math.Max(1, item.Quantity);
        if (item.Quantity <= 0) return 0;

        int hungerNeeded = player.MaxHunger - player.Hunger;
        int thirstNeeded = player.MaxThirst - player.Thirst;

        double hungerPerUnit = item.Nutrition.Hunger / (double)item.Quantity;
        double thirstPerUnit = item.Nutrition.Thirst / (double)item.Quantity;

        double amountForHunger = double.PositiveInfinity;
        double amountForThirst = double.PositiveInfinity;

        if (hungerPerUnit > 0 && hungerNeeded > 0)
            amountForHunger = Math.Ceiling(hungerNeeded / hungerPerUnit);
        if (thirstPerUnit > 0 && thirstNeeded > 0)
            amountForThirst = Math.Ceiling(thirstNeeded / thirstPerUnit);

        double optimal = Math.Min(item.Quantity, Math.Min(amountForHunger, amountForThirst));
        if (double.IsInfinity(optimal) || optimal <= 0)
            return item.Quantity;

        return (int)optimal;
    }

    /// <summary>
    /// Eat or drink <paramref name="item"/>. Partially consumes if
    /// <paramref name="amount"/> is given; otherwise auto-picks the optimal
    /// amount to fill hunger/thirst. If the item empties out, caller is
    /// responsible for removing the zero-quantity remnant from its pocket.
    /// Mirrors <c>consumeFood</c>.
    /// </summary>
    public static ConsumeResult ConsumeFood(Item item, Player player, int? amount = null)
    {
        if (item.Nutrition == null)
            return new ConsumeResult(false, "This is not consumable.", 0, 0, false);
        if (item.Quantity <= 0)
            return new ConsumeResult(false, "Nothing left to consume.", 0, 0, false);

        int consumeAmount;
        if (amount.HasValue) consumeAmount = amount.Value;
        else if (!string.IsNullOrEmpty(item.QuantityUnit))
            consumeAmount = CalculateOptimalConsumption(item, player);
        else
            consumeAmount = item.Quantity;

        if (consumeAmount <= 0) consumeAmount = 1;
        if (consumeAmount > item.Quantity)
            return new ConsumeResult(false, "Not enough to consume.", 0, item.Quantity, false);

        double ratio = item.Quantity > 0 ? (consumeAmount / (double)item.Quantity) : 1.0;
        int gainHunger = (int)Math.Round(item.Nutrition.Hunger * ratio);
        int gainThirst = (int)Math.Round(item.Nutrition.Thirst * ratio);

        if (gainHunger != 0)
            player.Hunger = Math.Clamp(player.Hunger + gainHunger, 0, player.MaxHunger);
        if (gainThirst != 0)
            player.Thirst = Math.Clamp(player.Thirst + gainThirst, 0, player.MaxThirst);

        bool contaminated = item.State.Contaminated;
        // Phase 3 will wire sickness as a status effect.  For now we just flag it.

        // Reduce quantity + proportional weight/volume
        int prevQty = item.Quantity;
        item.Quantity -= consumeAmount;
        if (prevQty > 0)
        {
            double wPerUnit = item.Weight / (double)prevQty;
            double vPerUnit = item.Volume / (double)prevQty;
            item.Weight = (int)Math.Floor(wPerUnit * item.Quantity);
            item.Volume = (int)Math.Floor(vPerUnit * item.Quantity);
        }

        string verb = item.Category == "drink" ? "drink" : "eat";
        string msg = contaminated
            ? $"You {verb} some of the {item.Name}. It tastes wrong."
            : $"You {verb} some of the {item.Name}.";

        return new ConsumeResult(true, msg, consumeAmount, item.Quantity, contaminated);
    }

    // ------------------------------------------------------------------
    // Per-turn world processing (spoilage + spillage)

    /// <summary>
    /// Advance contamination on opened food the player is carrying. Liquid-
    /// tagged spoils fastest (0.05/turn), protein slowest (0.03/turn), others
    /// 0.04/turn. Once ContaminationLevel &gt;= 0.3 the item flips contaminated.
    /// Mirrors <c>World.processFoodSpoilage</c>.
    /// </summary>
    public static void ProcessFoodSpoilage(Player player)
    {
        foreach (var item in player.Inventory.EnumerateReachableItems())
        {
            if (!item.IsContainer || !item.State.Opened) continue;
            foreach (var pocket in item.Pockets)
            {
                foreach (var content in pocket.Contents)
                {
                    if (content.Category != "food") continue;

                    double rate = 0.04;
                    if (content.HasTag("protein")) rate = 0.03;
                    if (content.HasTag("liquid"))  rate = 0.05;
                    content.State.ContaminationLevel += rate;
                    if (content.State.ContaminationLevel >= 0.3 && !content.State.Contaminated)
                        content.State.Contaminated = true;
                }
            }
        }
    }

    /// <summary>
    /// Spill 7 ml/turn from any opened, unsealed drink container the player
    /// is carrying. Liquids evaporate; stackable quantity drops.
    /// Mirrors <c>World.processLiquidSpillage</c>.
    /// </summary>
    public static void ProcessLiquidSpillage(Player player)
    {
        const int SPILL_PER_TURN = 7;  // ml
        foreach (var item in player.Inventory.EnumerateReachableItems())
        {
            if (!item.IsContainer || !item.State.Opened || item.State.Sealed) continue;
            foreach (var pocket in item.Pockets)
            {
                for (int i = pocket.Contents.Count - 1; i >= 0; i--)
                {
                    var content = pocket.Contents[i];
                    if (content.Category != "drink") continue;
                    if (content.Quantity <= 0 || content.QuantityUnit != "ml") continue;

                    int spill = Math.Min(SPILL_PER_TURN, content.Quantity);
                    int prevQty = content.Quantity;
                    content.Quantity -= spill;

                    double ratio = content.Quantity / (double)prevQty;
                    int prevW = content.Weight, prevV = content.Volume;
                    content.Weight = (int)Math.Floor(content.Weight * ratio);
                    content.Volume = (int)Math.Floor(content.Volume * ratio);
                    pocket.UsedWeight += content.Weight - prevW;
                    pocket.UsedVolume += content.Volume - prevV;

                    if (content.Quantity <= 0)
                    {
                        pocket.Remove(content);
                    }
                }
            }
        }
    }
}
